File: algorithm/_misc2/add-strings.cpp
// 415. Add Strings
// Difficulty: Easy

// Given two non-negative integers num1 and num2 represented as string, return the sum of num1 and num2.

// Note:

// The length of both num1 and num2 is < 5100.
// Both num1 and num2 contains only digits 0-9.
// Both num1 and num2 does not contain any leading zero.
// You must not use any built-in BigInteger library or convert the inputs to integer directly.

class Solution {
public:
  string addStrings(string num1, string num2) {
    string res;
    const int m = num1.length(), n = num2.length();

    size_t carry = 0;
    for (int i = 0; i < max(m, n); ++i) {
      const size_t a = i < m ? num1[m - 1 - i] - '0' : 0;
      const size_t b = i < n ? num2[n - 1 - i] - '0' : 0;
      size_t sum = carry + a + b;
      carry = sum / 10;
      res.push_back('0' + sum % 10);
    }
    if (carry) {
      res.push_back('0' + carry);
    }
    reverse(res.begin(), res.end());

    return res;
  }
};

File: data-structure-and-algorithm/javascript/leetcode/add-strings.js
/**
Given two non-negative integers, num1 and num2 represented as string, return the sum of num1 and num2 as a string.

You must solve the problem without using any built-in library for handling large integers (such as BigInteger). You must also not convert the inputs to integers directly.

 

Example 1:

Input: num1 = "11", num2 = "123"
Output: "134"
Example 2:

Input: num1 = "456", num2 = "77"
Output: "533"
Example 3:

Input: num1 = "0", num2 = "0"
Output: "0"
 

Constraints:

1 <= num1.length, num2.length <= 104
num1 and num2 consist of only digits.
num1 and num2 don't have any leading zeros except for the zero itself.
*/

/**
 * 
 * @param { string } num1 
 * @param { string } num2 
 * @returns { string }
 */
function addStrings(num1, num2) {
    let result = '';
    let carry = 0;
    const m = num1.length, n = num2.length;
    for (const i = 0; i < Math.max(m, n); ++i) {
        const a = i < m ? num1[m - 1 - i] - '0': 0;
        const b = i < n ? num2[n - 1 - i] - '0': 0;
        const sum = a + b + carry;
        carry = Math.trunc(sum / 10);
        result += sum % 10;
    }

    if (carry === 1) {
        result += carry;
    }

    return [...result].reverse().join("");
}